fhir_zulip_nlp/fhir_zulip_nlp.py

```diff
@@ -1,143 +1,208 @@
+    err_msg = None
+    messages: List[Dict] = []
+    try:
+        while True:
+            res: Dict = message_pull(anchor=anchor, num_before=0, num_after=num_after, keyword=keyword)
+            if 'retry-after' in res:  # rate limit, 200messages/user/minute
+                wait_seconds_required: float = res['retry-after']
+                time.sleep(wait_seconds_required * 1000)  # ms
+                continue
+            messages += res['messages']
+            if not messages:
+                print(f'No messages found for: {keyword}')
+                break
+            anchor = messages[-1]['id']  # returned messages are in chronological order; -1 is most recent in batch
+            # if i == 3:  # TODO: temp debugging
+            #     break
+            if res['found_newest']:  # this assumes API will reliably always return this
+                break
+
+    # TODO: @Rohan: We may never need this, but the API could change and things could break. I'm not sure whether or not
+    #  ...this is the best approach, continuing after error. Or maybe it should just exit? What do you think?
+    except Exception as err:
+        err_msg = str(err)
+        print(f'Got error querying {keyword}. Original error: {err}', file=sys.stderr)
+        print(f'Stopping for "{keyword}" and continuing on with the next keyword.', file=sys.stderr)
+
+    # TODO: Check: If a message contains the keyword more than once, will it return more than 1 result? or
+    #  ...simply 1 message result?
+    kw_report = {
+        'keyword': keyword,
+        'number of occurrences': len(messages),
+        # This is oldest/newest of all such messages. However, depending on caching / what was passed to this function,
+        # the oldest in `allmess` may be newer than the absolute oldest, and newest in `allmess` may be older than the
+        # absolute oldest.
+        'newest_datetime_this_session': format_timestamp(messages[0]['timestamp']) if messages else None,
+        'oldest_datetime_this_session': format_timestamp(messages[-1]['timestamp']) if messages else None,
+        'newest_datetime_absolute':
+            format_timestamp(message_pull('newest', 1, 0, keyword)['messages'][0]['timestamp']) if messages else None,
+        'oldest_datetime_absolute':
+            format_timestamp(message_pull('oldest', 0, 1, keyword)['messages'][0]['timestamp']) if messages else None,
+    }
+    if err_msg:
+        kw_report['error'] = str(err_msg)
+    return kw_report, messages
+
+
+def format_df(df: pd.DataFrame) -> pd.DataFrame:
+    """Format dataframe"""
+    # Reorganize columns
+    cols_head = ['category', 'keyword']
+    cols_tail = [x for x in list(df.columns) if x not in cols_head]
+    df = df[cols_head + cols_tail]
+    # Sort values
+    df = df.sort_values(cols_head + ['timestamp'] if 'timestamp' in cols_tail else [])
+    return df
+
+
+def query_categories(category_keywords: Dict[str, List[str]]) -> pd.DataFrame:
```

Comment:
@rohaher this function (previously `alldicts`) underwent a major refactor. sorry if it is a bit confusing. can probably simplify